### DISCUSSION TITLE
Does bpftrace support reading the local variable in uprobe?

### DISCUSSION THREAD

**Discussion by kingluo** [bpftrace/bpftrace#3010] [General]
It seems that bpftrace does not support DWARF, so how to read local variables?

A similar question is does it support file+offset uprobe?

**Comment by danobi** [comment]
bpftrace has some basic DWARF support; see: https://github.com/bpftrace/bpftrace/blob/master/docs/reference_guide.md#4-uprobeuretprobe-dynamic-tracing-user-level-arguments

But it currently does not support local variables. We could, in theory, enhance DWARF support to handle local vars, but it would require some work.

To manually work around the limitation, assuming compiler did not optimize them away, you could disassemble your function and figure out when the local var is live in a register and then access the register using `reg()`. Or figure out where on the stack it's stored and access it relative to stack pointer (assuming frame pointer is available).

**Reply by kingluo** [reply]
@danobi Thanks for your reply!
Another question is whether bpftrace supports setting uprobe at file+offset? It is more convenient than <function address>+offset, because setting probe at the source code rather than the assembly code is much more developer-friendly, doesn't it?

And I am also wondering if the `ustack()` is avaiable at all probes? Especially true for both uprobe and profile timer?

**Reply by danobi** [reply]
I presume you mean to ask if uprobe supports function+offset? In which case yes, bpftrace does support it: https://github.com/bpftrace/bpftrace/blob/master/docs/reference_guide.md#3-uprobeuretprobe-dynamic-tracing-user-level . Along with some additional safety features (so you can't accidentaly clobber the immediate value of insns by placing the probe).

IIUC ustack works when the probe interrupts userspace execution. For uprobe that is the case. For tracepoint it may, depending on if userspace is making a syscall, faulting a page, etc. Profile timer I suspect it would be hit or miss. Depending on what is currently running on CPU at the time.

**Comment by kingluo** [comment]
No, I mean c source file offset, like what systemtap supports.
Yes, I know the ustack must be filtered by current interrupted pid. But I'm just wondering if we can call ustack() in the profile timer probe type.



---- Replied Message ----
| From | Daniel ***@***.***> |
| Date | 02/24/2024 00:06 |
| To | ***@***.***> |
| Cc | jinhua ***@***.***>***@***.***> |
| Subject | Re: [bpftrace/bpftrace] Does bpftrace support reading the local variable in uprobe? (Discussion #3010) |

I presume you mean to ask if uprobe supports function+offset? In which case yes, bpftrace does support it: https://github.com/bpftrace/bpftrace/blob/master/docs/reference_guide.md#3-uprobeuretprobe-dynamic-tracing-user-level . Along with some additional safety features (so you can't accidentaly clobber the immediate value of insns by placing the probe).

IIUC ustack works when the probe interrupts userspace execution. For uprobe that is the case. For tracepoint it may, depending on if userspace is making a syscall, faulting a page, etc. Profile timer I suspect it would be hit or miss. Depending on what is currently running on CPU at the time.

**Reply by danobi** [reply]
Ah, ok. Not yet, but that would be a good feature. We have some basic DWARF support, so infrastructure is there. So adding file offset is definitely a possibility.

You are allowed to call ustack() in profile probe. It just may not always capture a stack:

```
dxu@fedora~ $ sudo bpftrace -e 'profile:s:1 { print(ustack) }' -f json
{"type": "attached_probes", "data": {"probes": 1}}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": "\n        __poll+77\n        0x7fa43026b526\n        pa_mainloop_poll+244\n        pa_mainloop_iterate+51\n        pa_mainloop_run+40\n        0x7fa43026f751\n        0x7fa42ffdbd7b\n        start_thread+759\n        __clone3+48\n"}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
{"type": "value", "data": ""}
^C
```

**Comment by kingluo** [comment]
Is it feasible to add a pid filter for profile timer probe, so that ustack() returns the valid stack which exactly belongs to the process we want?



---- Replied Message ----
| From | Daniel ***@***.***> |
| Date | 02/24/2024 00:31 |
| To | ***@***.***> |
| Cc | jinhua ***@***.***>***@***.***> |
| Subject | Re: [bpftrace/bpftrace] Does bpftrace support reading the local variable in uprobe? (Discussion #3010) |

Ah, ok. Not yet, but that would be a good feature. We have some basic DWARF support, so infrastructure is there. So adding file offset is definitely a possibility.

You are allowed to call ustack() in profile probe. It just may not always capture a stack:

***@***.***~ $ sudo bpftrace -e 'profile:s:1 { print(ustack) }'
[sudo] password for dxu:
Attaching 1 probe...















        0x7fda0374b8e0
        0x7fd9e7ff5eb8
        0x7fda036d3920
        0x841f0f







        0x7fda03910d1c


        g_main_context_prepare+921
        0x7fda4719e491
        g_main_loop_run+127
        meta_context_run_main_loop+90
        0x555dc233dfb7
        __libc_start_call_main+122
        __libc_start_main+139
        0x555dc233e295
